src/supertokens-auth.guard.ts

```diff
@@ -0,0 +1,249 @@
+import { CanActivate, ExecutionContext, Injectable } from "@nestjs/common"
+import { Reflector } from "@nestjs/core"
+import UserRoles from "supertokens-node/recipe/userroles"
+import {
+  getSession,
+  SessionClaimValidator,
+  VerifySessionOptions,
+} from "supertokens-node/recipe/session"
+import { Error as STError } from "supertokens-node/recipe/session"
+
+import { EmailVerificationClaim } from "supertokens-node/recipe/emailverification"
+import { MultiFactorAuthClaim } from "supertokens-node/recipe/multifactorauth"
+import { VerifySession, PublicAccess, Auth } from "./decorators"
+import { AuthDecoratorOptions, SuperTokensSession } from "./supertokens.types"
+import { JSONValue } from "supertokens-node/types"
+
+@Injectable()
+export class SuperTokensAuthGuard implements CanActivate {
+  private reflector: Reflector
+  constructor() {
+    this.reflector = new Reflector()
+  }
+
+  public async canActivate(context: ExecutionContext): Promise<boolean> {
+    // TODO: Exclude supertokens routes if the guard is applied globally
+    const isPublic = this.reflector.get(PublicAccess, context.getHandler())
+
+    if (isPublic) return true
+
+    const verifySessionOptions: VerifySessionOptions | undefined =
+      this.reflector.get(VerifySession, context.getHandler())
+    const authOptions: AuthDecoratorOptions | undefined = this.reflector.get(
+      Auth,
+      context.getHandler(),
+    )
+    const ctx = context.switchToHttp()
+    const req = ctx.getRequest()
+    const resp = ctx.getResponse()
+
+    let parsedVerifySessionOptions = verifySessionOptions
+    parsedVerifySessionOptions = this.addEmailVerificationToSessionOptions(
+      authOptions?.requireEmailVerification,
+      verifySessionOptions,
+    )
+    parsedVerifySessionOptions = this.removeMFAFromSessionOptions(
+      authOptions?.requireMFA,
+      verifySessionOptions,
+    )
+
+    const session = await getSession(req, resp, parsedVerifySessionOptions)
+    req.session = session
+
+    await this.validateRoles(session, authOptions?.roles)
+    await this.validatePermissions(session, authOptions?.permissions)
+    await this.validateMFA(session, authOptions?.requireMFA)
```

Comment:
I'd prefer if these used built-in validators and they were merged into the verify session options instead.

Reply:
I'm using the builting validators rn. But I'm not 100% sure that the MFA validation is the correct one.

Reply:
Hmm, looks like I was looking at the wrong version. The MFA validator is the correct one.